torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,79 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::optional<c10::ScalarType> typeToScalarType(const c10::TypePtr & type) {
+    if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    } else if (auto cast = type->cast<FloatType>()) {
+      return c10::ScalarType::Float;
+    } else if (auto cast = type->cast<IntType>()) {
+      return c10::ScalarType::Long;
+    } else if (auto cast = type->cast<BoolType>()) {
+      return c10::ScalarType::Bool;
+    }
+    return c10::nullopt;
+  }
+
+  // Promotes result types for arithmetic operations using new type promotion logic.
+  // See tensor_attributes.rst for details.
+  // This function handles the dimmed and zero-dim case, not the `wrapped-number`
+  // case. Wrapped number is handled separately because jit explicitly matches
+  // on the `Scalar other` signature.
+  c10::optional<c10::ScalarType> getPromotedTypeForArithmeticOp(Node *node) {
+    c10::ScalarType dimmed = c10::ScalarType::Undefined;
+    c10::ScalarType zerodim = c10::ScalarType::Undefined;
+    auto& args = node->schema().arguments();
+    // binary arithmetic ops, more than 2 args is alpha.
+    for (size_t i = 0 ; i < 2 ; i++ ) {
+      auto arg = args[i];
+      auto dtt = node->inputs()[i]->type()->cast<DimensionedTensorType>();
+      if ( dtt && dtt->dim() > 0) {
```

Comment:
This is only iterating over the first two args and the precondition before this function is called is that all inputs are DimensionedTensorType. I think it would make the code more readable if you established that invariant at the top of the function and from then on used : `node->inputs()[i]->type()->expect<DimensionedTensorType>()`, that way you don't have to repeatedly test for the `c10::nullopt` case.

Reply:
So what you're saying is that the nullopt case we need to handle can only occur for the `Scalar other` signature?

Reply:
It will only occur when the type isn't castable to DimensionedTensorType, (any type other than CompleteTensorType & DimensionedTensorType). This is being called only for 
```
            "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
            "aten::sub(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
            "aten::mul(Tensor self, Tensor other) -> Tensor",
            "aten::div(Tensor self, Tensor other) -> Tensor",
```
we're only iterating over the first two args, and the function is invoked only if all tensor args are castable to `DimensionedTensorType`

Reply:
OK, I (mistakenly) thought I could get any jit type here despite the signature. updated.